### PR TITLE
[8.x] Adds PHP 8.1 support to serializable closures

### PR DESCRIPTION
**Pull request by nunomaduro**
This pull request adds PHP 8.1 support to serializable closures by replacing **[opis/closure](https://github.com/opis/closure)** by **[laravel/serializable-closure](https://github.com/laravel/serializable-closure)**.

- We decided to fork the opis package as the upcoming version `4.x` is a complete rewrite on top of the FFI extension. As Laravel is a web framework, and FFI is not enabled by default in web requests (and some OS's), this fork allows us to keep using the 3.x series while adding support for new PHP versions - including PHP 8.1.
- Queued jobs, and cached routes, that were serialized with **[opis/closure](https://github.com/opis/closure)** will still work, as the package and the code still exists in Laravel `8.x`. Yet, the package **[opis/closure](https://github.com/opis/closure)** is scheduled for removal in Laravel `9.x`. By then, normally people no longer have queued jobs, and cached routes, using the old format. Of course, every new queued jobs, and cached routes, will be using **[laravel/serializable-closure](https://github.com/laravel/serializable-closure)** for serialization.
- Small caveat: Queued jobs, and cached routes, that used **[opis/closure](https://github.com/opis/closure)** for their serialized string formats won't work with PHP 8.1. Therefore, people need to ensure that all queued jobs, and cached routes, are already using the new **[laravel/serializable-closure](https://github.com/laravel/serializable-closure)** before migrating to PHP 8.1. One idea to speed up the process is having this pull request targeting `8.x` branch, instead of `php81-fixes`.

Remaining tasks before merging this pull request:

- [x] Review the code on **[laravel/serializable-closure](https://github.com/laravel/serializable-closure)**.
- [x] Review this pull request.
- [x] Decide if we target `8.x` branch, instead of `php81-fixes`, so people can start using these changes sooner.
- [x] Ask Nuno to tag a stable version of **[laravel/serializable-closure](https://github.com/laravel/serializable-closure)**.

### CODE REVIEW

File: src/Illuminate/Bus/Batch.php [hunk 1]
```diff
@@ -420,17 +419,15 @@ public function delete()
...
      * @param  \Throwable|null  $e
      * @return void
      */
     protected function invokeHandlerCallback($handler, Batch $batch, Throwable $e = null)
     {
         try {
-            return $handler instanceof SerializableClosure
-                ? $handler->__invoke($batch, $e)
-                : call_user_func($handler, $batch, $e);
+            return call_user_func($handler, $batch, $e);
```
Comment (GrahamCampbell) [inline src/Illuminate/Bus/Batch.php:430]: `$handler($batch, $e)`

File: src/Illuminate/Encryption/EncryptionServiceProvider.php [hunk 2]
```diff
@@ -37,8 +39,28 @@ protected function registerEncrypter()
...
      * @return void
+     *
+     * @deprecated It will be removed in Laravel 9.
      */
     protected function registerOpisSecurityKey()
+    {
+        if (\PHP_VERSION_ID < 80100) {
+            $config = $this->app->make('config')->get('app');
+
+            if (! class_exists(OpisSerializableClosure::class) || empty($config['key'])) {
```
Comment (GrahamCampbell) [inline src/Illuminate/Encryption/EncryptionServiceProvider.php:50]: no need for this class exists check. it will always exist

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/Illuminate/Encryption/EncryptionServiceProvider.php (hunk 2 above)
Comment (nunomaduro) [reply src/Illuminate/Encryption/EncryptionServiceProvider.php:50]: Well, `illuminate/encryption` may be used as isolated package. So I rather not touch that code.

File: src/Illuminate/Encryption/EncryptionServiceProvider.php [hunk 3]
```diff
@@ -37,8 +39,28 @@ protected function registerEncrypter()
      * Configure Opis Closure signing for security.
      *
      * @return void
+     *
+     * @deprecated It will be removed in Laravel 9.
```
Comment (driesvints) [inline src/Illuminate/Encryption/EncryptionServiceProvider.php:43]:
```suggestion
     * @deprecated Will be removed in a future Laravel version.
```

**Review by GrahamCampbell — APPROVED** [review]
👍 assuming this has been tested on a real app. :)

**Comment by nunomaduro** [comment]
Please create a new issue on the serializable closure repository.

**Pull request merged**